File: Python_Books/算法图解/grokking_algorithms_COPY/08_greedy_algorithms/javascript/01_set_covering.js
'use strict';

// You pass an array in, and it gets converted to a set.
let states_needed = new Set(["mt", "wa", "or", "id", "nv", "ut", "ca", "az"]);

const stations = {};
stations["kone"] = new Set(["id", "nv", "ut"]);
stations["ktwo"] = new Set(["wa", "id", "mt"]);
stations["kthree"] = new Set(["or", "nv", "ca"]);
stations["kfour"] = new Set(["nv", "ut"]);
stations["kfive"] = new Set(["ca", "az"]);

const final_stations = new Set();


while (states_needed.size) {
  let best_station = null;
  let states_covered = new Set();
  for (let station in stations) {
    let states = stations[station];
    let covered = new Set([...states_needed].filter((x) => states.has(x)));
    if (covered.size > states_covered.size) {
      best_station = station;
      states_covered = covered;
    }
  }
  states_needed = new Set([...states_needed].filter((x) => !states_covered.has(x)));
  final_stations.add(best_station);
}

console.log(final_stations); // Set { 'kone', 'ktwo', 'kthree', 'kfive' }

File: Python_Books/算法图解/grokking_algorithms_COPY/08_greedy_algorithms/lua/01_set_covering.lua
-- Custom set module
require "set"

-- You pass an array in, and it gets converted to a set.
local states_needed = set.new({"mt", "wa", "or", "id", "nv", "ut", "ca", "az"})

local stations = {}
stations["kone"] = set.new({"id", "nv", "ut"})
stations["ktwo"] = set.new({"wa", "id", "mt"})
stations["kthree"] = set.new({"or", "nv", "ca"})
stations["kfour"] = set.new({"nv", "ut"})
stations["kfive"] = set.new({"ca", "az"})

local final_stations = set.new()

while next(states_needed) ~= nil do
    local best_station = nil
    local states_covered = set.new()
    for station, states in pairs(stations) do
        local covered = states_needed * states
        if covered:len() > states_covered:len() then
            best_station = station
            states_covered = covered
        end
    end

    states_needed = states_needed - states_covered
    final_stations:add(best_station)
end

print(final_stations)